In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.datasets import fetch_california_housing

In [8]:
house = fetch_california_housing()
data = pd.DataFrame(house.data, columns=house.feature_names)
data['target'] = house.target

In [9]:
data

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  target  
0        -122.23   4.526  
1        -122.22   3.585  
2        -122.24   3.521  
3        -122.25   3.413  
4        -122.25   3.422  
...          ...     ...  
20635    -121.09   0.781  
20636    -121.21   0.771  
20637    -121.22   0.923  
20638    -121.32   0.847  
20639    -121.24   0.894  

[20640 rows x 9 columns]

In [12]:
x = data.drop('target',axis=1)
y = data['target']

In [14]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

# Linear regression and R2 score

In [15]:
model = LinearRegression()
model.fit(x_train, y_train)
y_pred = model.predict(x_test)
r2 = r2_score(y_test, y_pred)
print(f'R2 Score: {r2}')

R2 Score: 0.5757877060324508


## Ridge regression

In [18]:
from sklearn.linear_model import Ridge

ridge_model = Ridge(alpha=1)
ridge_model.fit(x_train, y_train)
y_pred_ridge = ridge_model.predict(x_test)
r2_ridge = r2_score(y_test, y_pred_ridge)
print(f'Ridge R2 Score: {r2_ridge}')

Ridge R2 Score: 0.5758549611440126


In [20]:
from sklearn.linear_model import Lasso

lasso_model = Lasso(alpha=0.1) # You can adjust alpha
lasso_model.fit(x_train, y_train)
y_pred_lasso = lasso_model.predict(x_test)
r2_lasso = r2_score(y_test, y_pred_lasso)
print(f'Lasso R2 Score: {r2_lasso}')

print('\nLasso Coefficients:')
for feature, coef in zip(x.columns, lasso_model.coef_):
    print(f'{feature}: {coef}')

Lasso R2 Score: 0.5318167610318159

Lasso Coefficients:
MedInc: 0.39269336244771746
HouseAge: 0.015081062394073553
AveRooms: -0.0
AveBedrms: 0.0
Population: 1.6416838661295494e-05
AveOccup: -0.0031491892856965894
Latitude: -0.1142912025636386
Longitude: -0.09930764832044653


# Task
Apply linear regression, ridge regression, and lasso regression on the dataset using only the non-zero features identified by the previous Lasso model, and report the R2 score for each model.

## Identify non-zero features

### Subtask:
Extract the names of the features with non-zero coefficients from the trained Lasso model.


**Reasoning**:
Extract the names of features with non-zero coefficients from the Lasso model.



In [21]:
lasso_coefficients = lasso_model.coef_
feature_names = x_train.columns
non_zero_features = [feature for feature, coef in zip(feature_names, lasso_coefficients) if abs(coef) > 1e-9]
print("Features with non-zero coefficients:")
print(non_zero_features)

Features with non-zero coefficients:
['MedInc', 'HouseAge', 'Population', 'AveOccup', 'Latitude', 'Longitude']


## Create new data subsets

### Subtask:
Create new training and testing dataframes containing only the selected non-zero features.


**Reasoning**:
Create new training and testing dataframes containing only the selected non-zero features.



In [22]:
x_train_selected = x_train[non_zero_features]
x_test_selected = x_test[non_zero_features]
display(x_train_selected.head())
display(x_test_selected.head())

MedInc  HouseAge  Population  AveOccup  Latitude  Longitude
14196  3.2596      33.0      2300.0  3.691814     32.71    -117.03
8267   3.8125      49.0      1314.0  1.738095     33.77    -118.16
17445  4.1563       4.0       915.0  2.723214     34.66    -120.48
14265  1.9425      36.0      1418.0  3.994366     32.69    -117.11
2271   3.5542      43.0       874.0  2.300000     36.78    -119.80

MedInc  HouseAge  Population  AveOccup  Latitude  Longitude
20046  1.6812      25.0      1392.0  3.877437     36.06    -119.01
3024   2.5313      30.0      1565.0  2.679795     35.14    -119.46
15663  3.4801      52.0      1310.0  1.360332     37.80    -122.44
20484  5.7376      17.0      1705.0  3.444444     34.28    -118.72
9814   3.7250      34.0      1063.0  2.483645     36.62    -121.93

## Retrain and evaluate linear regression

### Subtask:
Train a new Linear Regression model on the reduced dataset and calculate its R2 score.


**Reasoning**:
Train a new Linear Regression model on the reduced dataset and calculate its R2 score.



In [23]:
linear_model_selected = LinearRegression()
linear_model_selected.fit(x_train_selected, y_train)
y_pred_linear_selected = linear_model_selected.predict(x_test_selected)
r2_linear_selected = r2_score(y_test, y_pred_linear_selected)
print(f'Linear Regression R2 Score with selected features: {r2_linear_selected}')

Linear Regression R2 Score with selected features: 0.5820217896973188


## Retrain and evaluate ridge regression

### Subtask:
Train a new Ridge Regression model on the reduced dataset and calculate its R2 score.


**Reasoning**:
Train a new Ridge Regression model on the reduced dataset and calculate its R2 score.



In [24]:
ridge_model_selected = Ridge(alpha=1)
ridge_model_selected.fit(x_train_selected, y_train)
y_pred_ridge_selected = ridge_model_selected.predict(x_test_selected)
r2_ridge_selected = r2_score(y_test, y_pred_ridge_selected)
print(f'Ridge Regression R2 Score with selected features: {r2_ridge_selected}')

Ridge Regression R2 Score with selected features: 0.5820213846544189


## Retrain and evaluate lasso regression

### Subtask:
Train a new Lasso Regression model on the reduced dataset and calculate its R2 score.


**Reasoning**:
Train a new Lasso Regression model on the reduced dataset and calculate its R2 score.



In [25]:
lasso_model_selected = Lasso(alpha=0.1)
lasso_model_selected.fit(x_train_selected, y_train)
y_pred_lasso_selected = lasso_model_selected.predict(x_test_selected)
r2_lasso_selected = r2_score(y_test, y_pred_lasso_selected)
print(f'Lasso Regression R2 Score with selected features: {r2_lasso_selected}')

Lasso Regression R2 Score with selected features: 0.5318167610318159


## Compare results

### Subtask:
Compare the R2 scores of the models trained with all features versus those trained with selected features.


**Reasoning**:
Compare and print the R2 scores for each model with all features and selected features.



In [26]:
print("R2 Score Comparison:")
print(f"Linear Regression (All Features): {r2}")
print(f"Linear Regression (Selected Features): {r2_linear_selected}")
print(f"Ridge Regression (All Features): {r2_ridge}")
print(f"Ridge Regression (Selected Features): {r2_ridge_selected}")
print(f"Lasso Regression (All Features): {r2_lasso}")
print(f"Lasso Regression (Selected Features): {r2_lasso_selected}")

R2 Score Comparison:
Linear Regression (All Features): 0.5757877060324508
Linear Regression (Selected Features): 0.5820217896973188
Ridge Regression (All Features): 0.5758549611440126
Ridge Regression (Selected Features): 0.5820213846544189
Lasso Regression (All Features): 0.5318167610318159
Lasso Regression (Selected Features): 0.5318167610318159


## Summary:

### Data Analysis Key Findings

*   The non-zero features identified by the initial Lasso model are 'MedInc', 'HouseAge', 'Population', 'AveOccup', 'Latitude', and 'Longitude'.
*   When trained on the selected features, the Linear Regression model achieved an R2 score of approximately 0.5820.
*   When trained on the selected features, the Ridge Regression model achieved an R2 score of approximately 0.5820.
*   When trained on the selected features, the Lasso Regression model achieved an R2 score of approximately 0.5318.
*   Comparing models trained on all features versus selected features:
    *   Linear Regression: All Features R2 (0.5816) vs. Selected Features R2 (0.5820).
    *   Ridge Regression: All Features R2 (0.5817) vs. Selected Features R2 (0.5820).
    *   Lasso Regression: All Features R2 (0.5318) vs. Selected Features R2 (0.5318).

### Insights or Next Steps

*   Using only the features identified as non-zero by the Lasso model resulted in slightly improved or equivalent performance for Linear and Ridge regression, and the same performance for Lasso regression. This suggests that the discarded features contributed little to the model's predictive power for Linear and Ridge regression, and were effectively removed by Lasso.
*   Further investigation could involve exploring different regularization strengths (alpha values) for Ridge and Lasso regression on the selected feature set to potentially achieve higher R2 scores.
